torch/_subclasses/meta_utils.py

```diff
@@ -269,6 +269,99 @@ def empty_create(inner_t, inner_src, symbolic_context=symbolic_context):
+                    )
+                ),
+                outer_size=outer_size,
+                outer_stride=outer_stride,
+            )
+
+        # Returns a fake-ified version of t with the same view relationship wrt the given
+        # fake base. For views involving subclasses, this performs view replay, while
+        # dense -> dense views can be simplified to an as_strided() call. Note that view
+        # replay involves handling any closed-over state present in the view func:
+        #   * Closed-over tensors are fake-ified
+        #   * Closed-over SymInts are made symbolic
+        def view_from_base(base, t, source=source, shape_env=shape_env):
+            # fake-ify t naively; view relationship isn't correct yet
+            (sizes, strides, storage_offset) = sym_sizes_strides_storage_offset(
+                t, source
+            )
+            if not is_traceable_wrapper_subclass(
+                t
+            ) and not is_traceable_wrapper_subclass(base):
+                # Dense -> Dense view case uses as_strided() to construct view relationship.
+                # TODO: Change this logic to use view replay for consistency?
+                # It's likely there is no view func available.
+                return base.as_strided(sizes, strides, storage_offset)
+
+            def symint_visitor_fn(s):
+                if shape_env is None:
+                    return s
+
+                from torch._dynamo.source import ConstantSource
+
+                # NB: The source is meaningless; the symbol here is expected to be simplified out.
+                sym_source = ConstantSource("view_func")
+                val = int(s)
+                symbol = shape_env.create_symbol(val, sym_source)
+                return shape_env.create_symintnode(symbol, hint=val, source=sym_source)
+
+            # Assume the only closed-over tensors encountered will be subclass inner tensors.
+            real_to_fake_mapping = {}
+            if is_traceable_wrapper_subclass(t):
+                fake_t = empty_create_subclass(
+                    t, outer_size=sizes, outer_stride=strides
+                )
+                attrs, _ = fake_t.__tensor_flatten__()
+                for attr in attrs:
+                    real_to_fake_mapping[getattr(t, attr)] = getattr(fake_t, attr)
+
+            def tensor_visitor_fn(t):
+                return real_to_fake_mapping.get(t, t)
+
+            # Replay the view, swapping out any non-symbolic SymInts or real tensors
+            # for symbolic SymInts or fake tensors.
+            fake_t = t._view_func_unsafe(base, symint_visitor_fn, tensor_visitor_fn)
+
+            # Ensure the output has symbolic shapes according to the symbolic context.
+            # NB: These asserts simplify out any symbolicized closed-over view func state.
+            assert fake_t.size() == sizes
+            assert fake_t.stride() == strides
+            assert fake_t.storage_offset() == storage_offset
```

Comment:
@ezyang I'm interested in your opinion on whether this is sufficient. The idea is that I don't want the symbols created during view replay to end up in guards (so I don't have to expose view func state via a public API to apply the guard). I *think* I can get away with:
1. Allocate symbols for outer sizes, strides, storage offset for the view tensor
2. Do view replay, allocating symbols for any closed-over SymInts (use a meaningless Source under the expectation that it will be simplified out)
3. Run these asserts. I think this will result in the final fake view metadata having the outer sizes, strides, storage offset we allocated earlier, and simplifying out any symbolicized SymInts from the `symint_visitor_fn`, both of which we want.

In theory (and I think in practice based on the tests I added?) the SymInts allocated during view replay should never need to be guarded on directly.

Reply:
ShapeEnv doesn't give a guarantee on which symbol is eliminated when you do `torch._check(s0 == s1)`, so at minimum, we need to do something about this.

The other situation I'm worried about is similar to https://github.com/pytorch/pytorch/pull/118405/files#r1501566454 where you have NT1 -> Dense -> NT2 -> Dense. You will end up with anonymous symbols for NT2 but there isn't really any reason to believe they will show up in NT1, they could be totally unconnected, and you end with symbols that aren't bound anywhere (should write a test for this case!)

I am also worried about the symbol not being guaranteed to show up exactly as is at the end. For example, suppose we have a weird view foobar(SymInt x) which given a tensor of size `(0, y,)`, produces a tensor `(0, x * y)`. Now in the output, you would end up testing something like `s0 * 12 == s1 * 12`, but multiplication is not injective so you wouldn't actually manage to eliminate it. This one I'm less worried about it because it seems pretty difficult to devise a view function that can take a SymInt but then doesn't actually produce that SymInt as is on the output. I'm just pointing out that we never actually required this on our view functions.

It seems reasonable to choke with NYI for the second and third cases. So you need to (1) assert that those symbols were actually eliminated, and (2) make sure we only ever eliminate the anonymous symbols. If you make a dedicated source for them, maybe we can consult it when deciding which variable to eliminate. You could test this by swapping the assert test to be `strides == fake_t.stride()`.

Reply:
@ezyang Okay good points! I created a dedicated `EphemeralSource` for these which is taken into account during symbol simplification. It errors when guarded on so this essentially asserts that it's eliminated. See #120948

> NT1 -> Dense -> NT2 -> Dense
I added some tests for this in the next PR in the stack that introduces NJT views.